webcrypto/key.go

```diff
@@ -37,43 +30,24 @@ type CryptoKey[H KeyHandle] struct {
 	// `SubtleCrypto.exportKey` and `SubtleCrypto.wrapKey` will fail.
 	Extractable bool `json:"extractable"`
 
-	// FIXME: should be private?
-	// Algorithm holds the algorithm for which this key can be used
-	// and any associated extra parameters.
-	//
-	// The value of this property is an object that is specific to the
-	// algorithm used to generate the key. Possible values should be castable
-	// to the following types:
-	//   - AESKeyGenParams
-	//   - RSAHashedKeyGenParams
-	//   - ECKeyGenParams
-	//   - HMACKeyGenParams
-	Algorithm interface{} `json:"algorithm"`
-
-	// FIXME: should be private?
-	// Usages indicates what can be done with the key.
+	// By the time we access the Algorithm field of CryptoKey, we
+	// generally already know what type of algorithm it is, and are
+	// really looking to access the specific attributes of that algorithm.
+	// Thus, the generic parameter type helps us manipulate the
+	// `CryptoKey` type without having to cast the `Algorithm` field.
+	Algorithm A `json:"algorithm"`
+
+	// Usages holds the key usages for which this key can be used.
 	Usages []CryptoKeyUsage `json:"usages"`
 
 	// handle is an internal slot, holding the underlying key data.
 	// See [specification](https://www.w3.org/TR/WebCryptoAPI/#dfnReturnLink-0).
-	//nolint:unused
-	handle H
-}
-
-// KeyGenerator is an interface that represents a cryptographic key generator.
-type KeyGenerator interface {
-	GenerateKey(rt *goja.Runtime, extractable bool, keyUsages []CryptoKeyUsage) (goja.Value, error)
-}
-
-// KeyHandle represents the underlying type of a key data handle.
-type KeyHandle interface {
-	[]byte | crypto.PrivateKey | crypto.PublicKey
+	handle interface{}
```

Comment:
```suggestion
	handle any
```

You're already using plenty of generics, so why not here? :blush:

(BTW, when will we have that generics workshop? :wink: :pray:)

Reply:
I've actually switched back to using `any`, even for the `algorithm` attribute. As I've found out in later PRs, you cannot access the goja runtime from the promise callback without creating a race condition. Yet, most operations that were made easier and more robust by generics (we already know that in that case, we expect a params object describing a Aes Ctr algorithm so we can already use that specific type for our crypto keys) did rely on the ability to use the `Runtime().ExportTo` operation; which in turn forced me to do everything twice, just for the sake of generics.

We might find a workaround at a later point in time, and I'd be keen to bring generics back, then. However, at this point, they were really just forcing the code to be more complicated than necessary 😞

Reply:
Also, I'd be keen to make that now infamous workshop 🎉 but I guess I'd need to put up some specific, concrete, use case for it: WebCrypto was close 🤞🏻